src/compute/src/logging/compute.rs

```diff
@@ -864,99 +863,96 @@ impl<A: Allocate> DemuxHandler<'_, '_, A> {
-        export_ids: Vec<GlobalId>,
-    ) -> Self;
+/// Logging state maintained for a compute collection.
+///
+/// This type is used to produce appropriate log events in response to changes of logged collection
+/// state, e.g. frontiers, and to produce cleanup events when a collection is dropped.
+pub struct CollectionLogging {
+    id: GlobalId,
+    logger: Logger,
+
+    logged_frontier: Option<Timestamp>,
+    logged_import_frontiers: BTreeMap<GlobalId, Timestamp>,
 }
 
-impl<G, C> LogImportFrontiers for StreamCore<G, C>
-where
-    G: Scope<Timestamp = Timestamp>,
-    C: Container,
-{
-    fn log_import_frontiers(
-        self,
+impl CollectionLogging {
+    /// Create new logging state for the identified collection and emit initial logging events.
+    pub fn new(
+        id: GlobalId,
         logger: Logger,
-        import_id: GlobalId,
-        export_ids: Vec<GlobalId>,
+        dataflow_index: usize,
+        import_ids: impl Iterator<Item = GlobalId>,
     ) -> Self {
-        // Using `RetractImportFrontiers` ensures that retraction events are logged even when the
-        // dataflow is dropped before the input advances to the empty frontier.
-        let mut retractions = RetractImportFrontiers {
-            logger: logger.clone(),
-            import_id,
-            export_ids: export_ids.clone(),
-            time: None,
-        };
+        logger.log(ComputeEvent::Export { id, dataflow_index });
 
-        self.inspect_container(move |event| {
-            let Err(frontier) = event else { return };
+        let mut self_ = Self {
+            id,
+            logger,
+            logged_frontier: None,
+            logged_import_frontiers: Default::default(),
+        };
 
-            retractions.log();
+        // Initialize frontier logging.
+        let initial_frontier = Some(Timestamp::MIN);
+        self_.set_frontier(initial_frontier);
+        import_ids.for_each(|id| self_.set_import_frontier(id, initial_frontier));
 
-            let Some(&time) = frontier.get(0) else { return };
-            for &export_id in export_ids.iter() {
-                logger.log(ComputeEvent::ImportFrontier {
-                    import_id,
-                    export_id,
-                    time,
-                    diff: 1,
-                });
-                retractions.time = Some(time);
-            }
-        })
+        self_
     }
-}
 
-impl<G, Tr> LogImportFrontiers for Arranged<G, Tr>
-where
-    G: Scope<Timestamp = Timestamp>,
-    Tr: TraceReader + Clone,
-{
-    fn log_import_frontiers(
-        mut self,
-        logger: Logger,
-        import_id: GlobalId,
-        export_ids: Vec<GlobalId>,
-    ) -> Self {
-        self.stream = self
-            .stream
-            .log_import_frontiers(logger, import_id, export_ids);
-        self
+    /// Set the collection frontier to the given new time and emit corresponding logging events.
+    pub fn set_frontier(&mut self, new_time: Option<Timestamp>) {
```

Comment:
Yeah I considered that and it turns out more boilerplate-y. Also slightly more expensive because an `Antichain<Timestamp>` contains an allocation and an `Option<Timestamp>` does not, though I doubt that matters here.